src/txmempool.cpp

```diff
@@ -625,7 +625,7 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
     uint64_t innerUsage = 0;
 
     CCoinsViewCache mempoolDuplicate(const_cast<CCoinsViewCache*>(pcoins));
-    const int64_t spendheight = GetSpendHeight(mempoolDuplicate);
+    const int64_t spendheight = WITH_LOCK(::cs_main, return g_chainman.m_blockman.GetSpendHeight(mempoolDuplicate));
```

Comment:
The lock ordering is to take `cs_main` then take `mempool.cs` so at first glance, this looks like a lock inversion since it's locking `cs_main` while already holding `mempool.cs`. In fact, whenever `mempool.check()` is called, we're already holding `cs_main` (see the two call sites in net_processing).

I think it'd be better to add an `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` annotation to `check()` and remove the re-entrant lock-taking here.

Reply:
I agree with all the points made, I wanted to keep it this way as it doesn't change the existing (fragile) behaviour, but just makes it more obvious. However, I do think it'd be good to add the annotation.

I opened a new PR for this: https://github.com/bitcoin/bitcoin/pull/20972

Reply:
Fixed!